EVAL.md

```diff
@@ -0,0 +1,49 @@
+# Evaluation guidelines
+
+These are the guidelines used by the Resources team to evaluate project
+submissions.
+
+## Hard requirements
+
+The project must fulfill these requirements or it will be rejected.
+
+- The main logic of the application must be written in Rust.
+
+- The source code must be public. Licensing terms are unimportant.
+
+## Bonus points
+
+> N.B. A project repository may contain more than one crate. In those cases, one
+> (binary) crate will be the application, and the other crates will be libraries
+> and / or tools. In the review consider only the crates that are compiled for
+> the target device. One of these crates will be the application ("application
+> code") and the rest will be libraries ("support code").
+
+Award points if
+
+- The project is unlike the projects currently on display.
+
+- The documentation includes instructions on how to build the program, or the
+  firmware can be built with just `cargo build`.
+
+- The repository has a CI setup.
+
+- The application compiles on stable.
+
+- The application code is free from `unsafe` code and all `unsafe` code has been
+  pushed to libraries that only expose a safe API.
+
+- Code that's not target specific, if any, has unit tests.
+
+- Support code, if any, is documented.
+
+## Penalties
+
+Subtract points if
+
+- The application contains potential soundness issues. For example, using
+  `mem::transmute`, unchecked creation of singletons and unchecked use of
+  `static mut` variables are huge red flags.
+
+- Explicit panicking (e.g. `unwrap`) is used instead of proper error handling.
```

Comment:
Uhm, what is an explicit panic? There're valid uses of `unwrap()` and in many cases there's simply no other way of properly handling an error other than to panic.

Reply:
I guess we could ask for `expect("bla bla")` instead of `unwrap()` which at least gives a reason for the panic, though there might be binary size concerns...

Reply:
> Uhm, what is an explicit panic?

`unwrap`, `expect`, etc.

Compare:

Explicit panicking

``` rust
fn on_next_packet(bytes: &[u8]) {
    let packet = Packet::parse(bytes).unwrap();

    // do stuff
}
```

vs

Proper error handling (robust code)

``` rust
fn on_next_packet(bytes: &[u8]) {
    if let Ok(packet) = Packet::parse(bytes) {
        // do stuff
    } else {
        // we got junk
        #[cfg(debug_assertions)] // if in `dev` mode
        error!("malformed packet of {} bytes", bytes.len()); // log an error message
    }
}
```

> There're valid uses of unwrap()

IMO, `unwrap` should only be used when [you know the code won't panic
at runtime][1]. For example, `Resut::<T, !>.unwrap` is OK because there's no
alternative API.

[1]: https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html#cases-in-which-you-have-more-information-than-the-compiler

> in many cases there's simply no other way of properly handling an error other than to panic.

IMO, jumping to a "fail-safe" state when unrecoverable I/O errors happen is preferable
over panicking. Panicking should only happen due to bugs (programmer errors /
contract violations), which should be fixed during development phase. Handling
both cases in the same way makes it hard to spot real bugs.

``` rust
let byte = Serial.read().unwrap_or_else(|| fail_safe());

fn fail_safe() -> ! {
    FATAL_LED.on();

    // turn off actuators, or
    // sound alarm, or
    // w/e makes sense for this application

    // halt system
    loop {
        interrupt::disable();
    }
}
```

---

I can reword this line as "Explicit panicking (e.g. `unwrap`) is used where
proper error handling is possible.", but the bottom line here is that explicit
panicking is a "code smell" and demands a closer inspection of the code.

Reply:
> IMO, jumping to a "fail-safe" state when unrecoverable I/O errors happen is preferable
over panicking. Panicking should only happen due to bugs (programmer errors /
contract violations), which should be fixed during development phase.

Sure a failsafe would be preferable but that may not always be available, also for some applications halting is a no-go and a reset is preferred instead. There're are other classes or errors which can happen for which they might not be a useful handling procedure like if there're pluggable components and those are not present but required for operation.

Everything here has a big dependency on the application and the hardware and IMHO there's neither a one fits all nor a good/bad classification possible. I'd rather encourage finding creative ways to avoid panicking rather than penalisation. Also instead of `unwrap`ping an irrelevant `Result` I'd rather suggest using `ok()`.